Google.Api.Gax/GaxServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,41 @@
+﻿/*
+ * Copyright 2022 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using Google.Api.Gax;
+using System;
+using System.Reflection;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Extension methods enabling dependency injection of Google API clients.
+    /// </summary>
+    public static class GaxServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Adds the given Google API client type as a singleton, configuring it with any dependencies injected (e.g. credentials,
+        /// gRPC adapter). The client type must be annotated with <see cref="GoogleApiClientAttribute"/>;
+        /// the implementation creates an instance of the indicated <see cref="IClientBuilder{TClient}"/>,
+        /// populates it from the service-provider, and builds it.
+        /// </summary>
+        /// <typeparam name="TClient"></typeparam>
+        /// <param name="services"></param>
+        /// <returns></returns>
+        public static IServiceCollection AddGoogleApiClient<TClient>(this IServiceCollection services) where TClient : class
+        {
+            // TODO: Error checking, efficiency etc.
+            // TODO: Should we actually inherit this so we can inject an "impl"?
+            var attribute = typeof(TClient).GetCustomAttribute<GoogleApiClientAttribute>(inherit: false);
+            var builder = (IClientBuilder<TClient>) Activator.CreateInstance(attribute.BuilderType);
+            return services.AddSingleton(provider =>
+            {
+                builder.PopulateFromServices(provider);
```

Comment:
Any reason for making the builder not depend on DI? I think it's cleaner to put almost all the logic in the client builder base rather than duplicating it once per client. Happy to discuss though.

Reply:
I actually think I'd be more at ease if this was protected internal so only us can call it. I think we had to override in each library to get the "internal", but the override would just call base.

The reasons I don't like to have this public method depending on DI on the builder.

- We'd be only depending on MS DI mechanism, anyone wanting to use another container would have to write their own logic. That's a non-issue if this method is by the MS DI extension methods instead of in the builder itself.
- On more conceptual grounds the builder really should be indepent of DI, I think DI is yet another way to configure the builder and DI should figure out which aspects of the builder to configure and when, instead of the builder trying to get "all you can" from DI. "All you can" is just one use case of how the builder can be configured through DI but not necessarily the only one.
- The method being public means that calling code can call it anytime (in particular from the configurator function I'm proposing on the extension method), but without having control on what the method does, i.e. what it fetches from DI. So calling code would have to take all of that into account, call this method in "the right order" and then override or initialize other aspects of the builder etc.

What I'm proposing also is that this method doen't blindly get everything it can from DI, but only stuff that has not yet been configured. If the client set a credential path on the configuration function then we don't try to get a credential from DI, etc. That's why it makes more sense to have it outside the builder (so as not to give the impression it may be called repeatedly) and more importantly not public.